WordPress/src/main/java/org/wordpress/android/ui/reader/adapters/ReaderSearchSuggestionAdapter.java

```diff
@@ -1,53 +1,58 @@
 package org.wordpress.android.ui.reader.adapters;
 
+import android.app.AlertDialog;
 import android.content.Context;
+import android.content.DialogInterface;
 import android.database.Cursor;
+import android.database.MatrixCursor;
 import android.support.v4.widget.CursorAdapter;
-import android.text.TextUtils;
 import android.view.LayoutInflater;
 import android.view.View;
 import android.view.ViewGroup;
-import android.widget.ImageView;
 import android.widget.TextView;
 
 import org.wordpress.android.R;
 import org.wordpress.android.datasets.ReaderSearchTable;
 
 public class ReaderSearchSuggestionAdapter extends CursorAdapter {
     private static final int MAX_SUGGESTIONS = 5;
+    private static final int CLEAR_ALL_ROW_ID = -1;
+
     private String mCurrentFilter;
 
+    private final String mClearAllText;
+    private final int mClearAllBgColor;
+
     public ReaderSearchSuggestionAdapter(Context context) {
         super(context, null, false);
+        mClearAllText = context.getString(R.string.label_clear_saved_searches);
+        mClearAllBgColor = context.getResources().getColor(R.color.grey_lighten_30);
     }
 
-    /*
-     * populate the adapter using a cursor containing past searches that match the filter
-     */
     public void setFilter(String filter) {
-        // skip if unchanged
-        if (isCurrentFilter(filter) && getCursor() != null) {
-            return;
-        }
-        Cursor cursor = ReaderSearchTable.getQueryStringCursor(filter, MAX_SUGGESTIONS);
-        swapCursor(cursor);
-        mCurrentFilter = filter;
-    }
+        // get db cursor containing matching query strings
+        Cursor sqlCursor = ReaderSearchTable.getQueryStringCursor(filter, MAX_SUGGESTIONS);
 
-    /*
-     * forces setFilter() to always repopulate by skipping the isCurrentFilter() check
-     */
-    private void reload() {
-        String newFilter = mCurrentFilter;
-        mCurrentFilter = null;
-        setFilter(newFilter);
-    }
+        // create a MatrixCursor which will be the actual cursor behind this adapter
+        MatrixCursor matrixCursor = new MatrixCursor(
+                new String[]{
+                        ReaderSearchTable.COL_ID,
+                        ReaderSearchTable.COL_QUERY});
 
-    private boolean isCurrentFilter(String filter) {
-        if (TextUtils.isEmpty(filter) && TextUtils.isEmpty(mCurrentFilter)) {
-            return true;
+        if (sqlCursor.moveToFirst()) {
+            // first populate the matrix from the db cursor...
+            do {
+                long id = sqlCursor.getLong(sqlCursor.getColumnIndex(ReaderSearchTable.COL_ID));
+                String query = sqlCursor.getString(sqlCursor.getColumnIndex(ReaderSearchTable.COL_QUERY));
+                matrixCursor.addRow(new Object[]{id, query});
+            } while (sqlCursor.moveToNext());
+
+            // ...then add our custom item
+            matrixCursor.addRow(new Object[]{CLEAR_ALL_ROW_ID, mClearAllText});
```

Comment:
Rather than storing `mClearAllText` we could store the row data and prevent new `Object[]` allocations as the user types. ie remove `mClearAllText` and create `Object[] mClearRow` in the constructor.